src/nvim/msgpack_rpc/channel.c

```diff
@@ -359,21 +359,17 @@ static void on_request_event(void **argv)
   Channel *channel = e->channel;
   MsgpackRpcRequestHandler handler = e->handler;
   Array args = e->args;
-  uint64_t request_id = e->request_id;
+  uint32_t request_id = e->request_id;
   Error error = ERROR_INIT;
   Object result = handler.fn(channel->id, args, &error);
-  if (request_id != NO_RESPONSE) {
```

Comment:
This `if` is needed.

Reply:
Why? I do not think so.

First of all, there must not be exclusively allocated `request_id` or `response_id` since it is just a way to link a response and a request. [Here](https://github.com/neovim/neovim/blob/master/src/nvim/msgpack_rpc/channel.c#L295) `request_id` is initialized in [`msgpack_rpc_validate()`](https://github.com/neovim/neovim/blob/master/src/nvim/msgpack_rpc/helpers.c#L570). It is default value is `NO_RESPONSE`. Generally, this is slightly incorrect since `request_id` is required by [`serialize_response()`](https://github.com/neovim/neovim/blob/master/src/nvim/msgpack_rpc/channel.c#L300) if validation failed. However, it could be failed due to wrong type of message (for example, array have one element instead of four elements). So, there is no valid `request_id`. If validation is successful then `request_id` value is valid.

The second one reason to remove this `if` is that `request_id` is chosen by client. If client sends `UINT32_MAX - 1` messages then this `if` fails correct request.

Fortunately, `request_id` value is valid bellow [this error handling `if`](https://github.com/neovim/neovim/blob/master/src/nvim/msgpack_rpc/channel.c#L315).

Reply:
This makes sense to me. @bfredl still objection?

Reply:
`if (ERROR_SET(&error))` (line 297) handles the case where `msgpack_rpc_validate` fails, so this condition should never be true. We can return a bool from `msgpack_rpc_validate` and assert it, to be sure.

`serialize_response` is the only other case that checks `NO_RESPONSE`, but that's also unnecessary, it should check `err` or some bool parameter instead.

Reply:
Ah, but `NO_RESPONSE` is used for notifications ...

Reply:
as @bfredl said, this check is needed because the message-type may be "notification" which must not send a response. 

Much of the naming is confusing so I tried to improve it in https://github.com/neovim/neovim/pull/9887 . That also eliminates NO_RESPONSE.